### DISCUSSION TITLE
Preserve circular geometries when refining mesh

### DISCUSSION THREAD

**Discussion by TheGreatCid** [idaholab/moose#24846] [Q&A General] [answered]
Hi all,

I am making a cylindrical mesh using meshing tools built into moose. I start with a coarse mesh and then refine it as shown below:
```
  [gmg1]
    type = AnnularMeshGenerator
    nt = 40
    rmax = 5.13
    rmin = 4.75
    nr = 2
  []
  [extrude1]
    type = AdvancedExtruderGenerator
    input = gmg1
    heights = '1.225 0.05 1.225'
    num_layers = '2 5 2'
    direction = '0 0 1'
  []
  [sub]
    type = SubdomainBoundingBoxGenerator
    block_id = 5
    block_name = 'ref'
    input = extrude1
    bottom_left = '-1.25 4.63 -1'
    top_right = '1.25 5.63 10'
  []
  [refine]
    type = RefineBlockGenerator
    input = sub
    block = 5
    refinement = 2
  []
```
![image](https://github.com/idaholab/moose/assets/75846521/5c021358-8dab-46b9-b41c-fdbc90789aec)

This produces a cylindrical mesh. However, the issue is that the refinement interpolates points between nodes without knowledge of the original circular geometry, resulting in flat surfaces after refinement. I'm curious if there is a workaround for this or if I will need to try to generate this outside of MOOSE

**Comment by GiudGiud** [comment]
Hello

We have this generator to do that, in 2D
Then you can extrude the corrected boundary to 3D
https://mooseframework.inl.gov/source/meshgenerators/CircularBoundaryCorrectionGenerator.html

Guillaume

**Reply by TheGreatCid** [reply]
The issue for me is that I am looking for the refinement pattern pictured above with hanging nodes. I don't believe I can accomplish that with this method. I would need to refine the 2D mesh locally, apply the circular correction, and then extrude it, which would make a uniform amount of layers around the circumference of the domain, which I am trying to avoid due to computational costs.

**Reply by GiudGiud** [reply]
` uniform amount of layers around the circumference of the domain`
that does not have to be. Your 2D mesh does not need to have that, and you dont need to refine it uniformly around the circumference either

**Reply by TheGreatCid** [reply]
I could be using the extruder wrong because I don't understand how to achieve that. For  `AdvancedExtruderGenerator` the only variability you have is the number of layers between certain heights using the `heights` and `num_layers` argument. I don't have the option to say, let the part of the mesh within a bounding box extrude with 10 layers and the rest of the mesh extrude with only 5 layers, which is what I would need to achieve something similar to the refinement pictured in original post

**Reply by GiudGiud** [reply, marked as answer]
oh right.

Yeah can't do that.

You could adapt the mesh generator code for 3D

**Reply by TheGreatCid** [reply]
yea, I'll take a look at the circular correction code and see how easy it would be to modify
